src/serial.rs

```diff
@@ -0,0 +1,538 @@
+//! UART peripheral control
+//!
+//! Controls the 3 uart peripherals (UART0, UART1, UART2)
+//!
+//! **It currently depends on GPIO pins and clock to be configured with default settings.**
+//! (Tested for UART 0)
+//!
+//! # TODO
+//! - Automatic GPIO configuration
+//! - Add all extra features esp32 supports (eg rs485, etc. etc.)
+//! - Free APB lock when TX is idle (and no RX used)
+
+use core::convert::Infallible;
+use core::marker::PhantomData;
+
+use embedded_hal::serial;
+
+use crate::esp32::{UART0, UART1, UART2};
+use crate::units::*;
+
+const UART_FIFO_SIZE: u8 = 128;
+
+/// Serial error
+#[derive(Debug)]
+pub enum Error {
+    /// Framing error
+    Framing,
+    /// Noise error
+    Noise,
+    /// RX buffer overrun
+    Overrun,
+    /// Parity check error
+    Parity,
+    /// Baudrate too low
+    BaudrateTooLow,
+    /// Baudrate too high
+    BaudrateTooHigh,
+}
+
+/// Interrupt event
+pub enum Event {
+    /// New data has been received
+    Rxne,
+    /// New data can be sent
+    Txe,
+    /// Idle line state detected
+    Idle,
+}
+
+pub mod config {
+    use crate::units::*;
+
+    #[derive(PartialEq, Eq, Copy, Clone, Debug)]
+    pub enum DataBits {
+        DataBits5,
+        DataBits6,
+        DataBits7,
+        DataBits8,
+    }
+
+    #[derive(PartialEq, Eq, Copy, Clone, Debug)]
+    pub enum Parity {
+        ParityNone,
+        ParityEven,
+        ParityOdd,
+    }
+
+    #[derive(PartialEq, Eq, Copy, Clone, Debug)]
+    pub enum StopBits {
+        #[doc = "1 stop bit"]
```

Comment:
```suggestion
        /// 1 stop bit
```

Same on other fields, it's more ergonomic to use doc comments :wink:

Reply:
Will change indeed (I copied this from stm32)

Reply:
Done.